lib/grandpa/finalization.go

```
@@ -0,0 +1,493 @@
+// services and return an error, this function also returns if the
+// finalizationHandler.Stop() method is called
+func (fh *finalizationHandler) waitServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		finalizationEngineErr <- fh.finalizationEngine.Run()
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		votingRoundErr <- fh.votingRound.Run()
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err := <-votingRoundErr:
+			if err == nil {
+				votingRoundErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+
+		case err := <-finalizationEngineErr:
+			if err == nil {
+				finalizationEngineErr = nil
+				// go out from the select case
+				break
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Warnf("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+		}
+
+		finish := votingRoundErr == nil && finalizationEngineErr == nil
+		if finish {
+			return nil
+		}
+	}
+}
+
+// handleVotingRound interacts with finalizationEngine service
+// executing the actions based on what it receives throuhg channel
+type handleVotingRound struct {
+	grandpaService       *Service
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	if h.engineDone == nil {
+		return nil
+	}
+
```

imo I think we should not check this and let it panic... we shouldn't stop that service twice anyway right?

---

You're right, this check is not necessary!